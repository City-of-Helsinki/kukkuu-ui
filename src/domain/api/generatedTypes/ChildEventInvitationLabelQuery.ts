/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: ChildEventInvitationLabelQuery
// ====================================================

export interface ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node_EventGroupNode {
  /**
   * The ID of the object.
   */
  id: string;
  canChildEnroll: boolean | null;
}

export interface ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node_EventNode {
  /**
   * The ID of the object.
   */
  id: string;
  canChildEnroll: boolean | null;
}

export type ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node =

    | ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node_EventGroupNode
    | ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node_EventNode;

export interface ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges {
  /**
   * The item at the end of the edge
   */
  node: ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges_node | null;
}

export interface ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups {
  /**
   * Contains the nodes in this connection.
   */
  edges: (ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups_edges | null)[];
}

export interface ChildEventInvitationLabelQuery_child {
  /**
   * The ID of the object.
   */
  id: string;
  /**
   * All upcoming events and event groups for the child's project.
   */
  upcomingEventsAndEventGroups: ChildEventInvitationLabelQuery_child_upcomingEventsAndEventGroups | null;
}

export interface ChildEventInvitationLabelQuery {
  child: ChildEventInvitationLabelQuery_child | null;
}

export interface ChildEventInvitationLabelQueryVariables {
  childId: string;
}
