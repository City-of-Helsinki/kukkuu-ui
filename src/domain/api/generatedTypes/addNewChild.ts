/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { AddChildMutationInput } from './globalTypes';

// ====================================================
// GraphQL mutation operation: addNewChild
// ====================================================

export interface addNewChild_addChild_child_project {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  year: number;
}

export interface addNewChild_addChild_child {
  /**
   * The ID of the object.
   */
  id: string;
  firstName: string;
  lastName: string;
  birthdate: any;
  postalCode: string;
  project: addNewChild_addChild_child_project;
}

export interface addNewChild_addChild {
  child: addNewChild_addChild_child | null;
}

export interface addNewChild {
  /**
   * This mutation cannot be used before one has started using the service with "SubmitChildrenAndGuardianMutation".
   */
  addChild: addNewChild_addChild | null;
}

export interface addNewChildVariables {
  input: AddChildMutationInput;
}
