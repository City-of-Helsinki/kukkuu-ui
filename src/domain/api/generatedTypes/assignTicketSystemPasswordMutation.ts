/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { AssignTicketSystemPasswordMutationInput } from "./globalTypes";

// ====================================================
// GraphQL mutation operation: assignTicketSystemPasswordMutation
// ====================================================

export interface assignTicketSystemPasswordMutation_assignTicketSystemPassword {
  /**
   * The assigned ticket system password
   */
  password: string | null;
}

export interface assignTicketSystemPasswordMutation {
  assignTicketSystemPassword: assignTicketSystemPasswordMutation_assignTicketSystemPassword | null;
}

export interface assignTicketSystemPasswordMutationVariables {
  input: AssignTicketSystemPasswordMutationInput;
}
