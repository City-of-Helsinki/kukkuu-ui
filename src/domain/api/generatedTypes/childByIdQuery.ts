/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { EventParticipantsPerInvite, RelationshipTypeEnum } from "./globalTypes";

// ====================================================
// GraphQL query operation: childByIdQuery
// ====================================================

export interface childByIdQuery_child_project {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  year: number;
}

export interface childByIdQuery_child_occurrences_edges_node_venue {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  address: string | null;
}

export interface childByIdQuery_child_occurrences_edges_node_event {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  shortDescription: string | null;
  /**
   * In minutes
   */
  duration: number | null;
  image: string;
  imageAltText: string | null;
  participantsPerInvite: EventParticipantsPerInvite;
}

export interface childByIdQuery_child_occurrences_edges_node_enrolments_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  /**
   * An unique encoded reference id
   */
  referenceId: string | null;
}

export interface childByIdQuery_child_occurrences_edges_node_enrolments_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_occurrences_edges_node_enrolments_edges_node | null;
}

export interface childByIdQuery_child_occurrences_edges_node_enrolments {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_occurrences_edges_node_enrolments_edges | null)[];
}

export interface childByIdQuery_child_occurrences_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  time: any;
  venue: childByIdQuery_child_occurrences_edges_node_venue;
  event: childByIdQuery_child_occurrences_edges_node_event;
  enrolments: childByIdQuery_child_occurrences_edges_node_enrolments;
}

export interface childByIdQuery_child_occurrences_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_occurrences_edges_node | null;
}

export interface childByIdQuery_child_occurrences {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_occurrences_edges | null)[];
}

export interface childByIdQuery_child_availableEventsAndEventGroups_edges_node_EventNode {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  shortDescription: string | null;
  image: string;
  imageAltText: string | null;
  participantsPerInvite: EventParticipantsPerInvite;
}

export interface childByIdQuery_child_availableEventsAndEventGroups_edges_node_EventGroupNode {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  shortDescription: string | null;
  image: string;
  imageAltText: string | null;
}

export type childByIdQuery_child_availableEventsAndEventGroups_edges_node = childByIdQuery_child_availableEventsAndEventGroups_edges_node_EventNode | childByIdQuery_child_availableEventsAndEventGroups_edges_node_EventGroupNode;

export interface childByIdQuery_child_availableEventsAndEventGroups_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_availableEventsAndEventGroups_edges_node | null;
}

export interface childByIdQuery_child_availableEventsAndEventGroups {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_availableEventsAndEventGroups_edges | null)[];
}

export interface childByIdQuery_child_pastEvents_edges_node_occurrences_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  time: any;
}

export interface childByIdQuery_child_pastEvents_edges_node_occurrences_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_pastEvents_edges_node_occurrences_edges_node | null;
}

export interface childByIdQuery_child_pastEvents_edges_node_occurrences {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_pastEvents_edges_node_occurrences_edges | null)[];
}

export interface childByIdQuery_child_pastEvents_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  shortDescription: string | null;
  image: string;
  imageAltText: string | null;
  participantsPerInvite: EventParticipantsPerInvite;
  occurrences: childByIdQuery_child_pastEvents_edges_node_occurrences;
}

export interface childByIdQuery_child_pastEvents_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_pastEvents_edges_node | null;
}

export interface childByIdQuery_child_pastEvents {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_pastEvents_edges | null)[];
}

export interface childByIdQuery_child_relationships_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  type: RelationshipTypeEnum | null;
}

export interface childByIdQuery_child_relationships_edges {
  /**
   * The item at the end of the edge
   */
  node: childByIdQuery_child_relationships_edges_node | null;
}

export interface childByIdQuery_child_relationships {
  /**
   * Contains the nodes in this connection.
   */
  edges: (childByIdQuery_child_relationships_edges | null)[];
}

export interface childByIdQuery_child {
  /**
   * The ID of the object.
   */
  id: string;
  firstName: string;
  lastName: string;
  birthdate: any;
  postalCode: string;
  project: childByIdQuery_child_project;
  occurrences: childByIdQuery_child_occurrences;
  availableEventsAndEventGroups: childByIdQuery_child_availableEventsAndEventGroups | null;
  pastEvents: childByIdQuery_child_pastEvents | null;
  relationships: childByIdQuery_child_relationships;
}

export interface childByIdQuery {
  /**
   * The ID of the object
   */
  child: childByIdQuery_child | null;
}

export interface childByIdQueryVariables {
  id: string;
}
