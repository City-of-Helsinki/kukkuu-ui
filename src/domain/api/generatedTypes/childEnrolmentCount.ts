/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: ChildEnrolmentCount
// ====================================================

export interface ChildEnrolmentCount_child_project {
  /**
   * The ID of the object.
   */
  id: string;
  /**
   * How many times a single user can participate events per year. Changing this will not affect any existing enrolments.
   */
  enrolmentLimit: number;
}

export interface ChildEnrolmentCount_child {
  /**
   * The ID of the object.
   */
  id: string;
  /**
   * How many past enrolments child has this year.
   */
  pastEnrolmentCount: number | null;
  project: ChildEnrolmentCount_child_project;
}

export interface ChildEnrolmentCount {
  child: ChildEnrolmentCount_child | null;
}

export interface ChildEnrolmentCountVariables {
  childId: string;
}
