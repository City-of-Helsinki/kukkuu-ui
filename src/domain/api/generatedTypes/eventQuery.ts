/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

import { EventParticipantsPerInvite } from "./globalTypes";

// ====================================================
// GraphQL query operation: eventQuery
// ====================================================

export interface eventQuery_event_occurrences_edges_node_event {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
}

export interface eventQuery_event_occurrences_edges_node_venue {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  address: string | null;
}

export interface eventQuery_event_occurrences_edges_node {
  /**
   * The ID of the object.
   */
  id: string;
  time: any;
  remainingCapacity: number | null;
  event: eventQuery_event_occurrences_edges_node_event;
  venue: eventQuery_event_occurrences_edges_node_venue;
  childHasFreeSpotNotificationSubscription: boolean | null;
}

export interface eventQuery_event_occurrences_edges {
  /**
   * The item at the end of the edge
   */
  node: eventQuery_event_occurrences_edges_node | null;
}

export interface eventQuery_event_occurrences {
  /**
   * Contains the nodes in this connection.
   */
  edges: (eventQuery_event_occurrences_edges | null)[];
}

export interface eventQuery_event {
  /**
   * The ID of the object.
   */
  id: string;
  name: string | null;
  description: string | null;
  shortDescription: string | null;
  image: string;
  imageAltText: string | null;
  participantsPerInvite: EventParticipantsPerInvite;
  /**
   * In minutes
   */
  duration: number | null;
  capacityPerOccurrence: number;
  occurrences: eventQuery_event_occurrences;
}

export interface eventQuery {
  /**
   * The ID of the object
   */
  event: eventQuery_event | null;
}

export interface eventQueryVariables {
  id: string;
  date?: any | null;
  time?: any | null;
  childId?: string | null;
}
