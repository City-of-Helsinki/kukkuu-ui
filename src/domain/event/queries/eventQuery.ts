import gql from 'graphql-tag';

// TODO: Add remainingCapacity to occurrences when backend bug is fixed
const eventQuery = gql`
  query eventQuery($id: ID!) {
    event(id: $id) {
      id
      name
      description
      shortDescription
      image
      participantsPerInvite
      duration
      capacityPerOccurrence
      occurrences {
        edges {
          node {
            id
            time
            venue {
              name
              description
              address
            }
          }
        }
      }
    }
  }
`;
export default eventQuery;
