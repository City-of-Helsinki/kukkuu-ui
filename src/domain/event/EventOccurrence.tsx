import { useTranslation } from 'react-i18next';
import { useParams } from 'react-router-dom';

import { TicketSystem } from '../api/generatedTypes/globalTypes';
import { eventQuery_event_occurrences_edges_node as OccurrencesEdgeNode } from '../api/generatedTypes/eventQuery';
import { formatTime, newMoment } from '../../common/time/utils';
import LinkButton from '../../common/components/button/LinkButton';
import {
  DEFAULT_TIME_FORMAT,
  DEFAULT_DATE_FORMAT,
} from '../../common/time/TimeConstants';
import EventOccurrenceNotificationControlButton from './EventOccurrenceNotificationControlButton';
import styles from './eventOccurrence.module.scss';

const SubmitTypes = {
  enrol: 'ENROL',
  notification: 'NOTIFICATION',
  externalTicketSystem: 'EXTERNAL_TICKET_SYSTEM',
} as const;

type SubmitType = (typeof SubmitTypes)[keyof typeof SubmitTypes];

function getSubmitType(
  isExternalTicketSystem: boolean,
  hasCapacity: boolean
): SubmitType {
  if (isExternalTicketSystem) {
    return SubmitTypes.externalTicketSystem;
  }

  if (hasCapacity) {
    return SubmitTypes.enrol;
  }

  return SubmitTypes.notification;
}

type UrlParams = {
  childId: string;
  eventId: string;
};

type EventOccurrenceProps = {
  occurrence: OccurrencesEdgeNode;
  showFreePlaces?: boolean;
  canEnroll?: boolean | null;
};

const EventOccurrence = ({
  occurrence,
  showFreePlaces = true,
  canEnroll = true,
}: EventOccurrenceProps) => {
  const { t } = useTranslation();
  const { childId, eventId } = useParams<UrlParams>();

  const date = formatTime(
    newMoment(occurrence.time),
    `ddd ${DEFAULT_DATE_FORMAT}`
  );
  const time = formatTime(newMoment(occurrence.time), DEFAULT_TIME_FORMAT);
  const machineReadableDate = formatTime(newMoment(occurrence.time));
  const machineReadableTime = formatTime(
    newMoment(occurrence.time),
    'HH:mm:ss.SSS'
  );
  const machineReadableDateTime = `${machineReadableDate} ${machineReadableTime}`;

  const hasCapacity = Boolean(
    occurrence.remainingCapacity && occurrence.remainingCapacity > 0
  );
  const childHasSubscription = Boolean(
    occurrence.childHasFreeSpotNotificationSubscription
  );
  const remainingCapacity = hasCapacity
    ? occurrence?.remainingCapacity
    : t('event.register.occurrenceTableBody.full');
  const eventUrl = `${occurrence.event.id}/redirect`;
  const occurrenceUrl = `occurrence/${occurrence.id}/enrol`;
  const isExternalTicketSystem = [
    TicketSystem.TICKETMASTER,
    TicketSystem.LIPPUPISTE,
  ].includes(occurrence?.ticketSystem?.type as TicketSystem);
  const submitType = getSubmitType(isExternalTicketSystem, hasCapacity);
  const submitCell = (
    <>
      {submitType === SubmitTypes.enrol && (
        <LinkButton to={occurrenceUrl}>
          {t('event.register.occurrenceTableHeader.buttonText')}
        </LinkButton>
      )}
      {submitType === SubmitTypes.notification && (
        <EventOccurrenceNotificationControlButton
          childId={childId ?? ''}
          eventId={eventId ?? ''}
          isSubscribed={childHasSubscription}
          occurrence={occurrence}
          unsubscribeLabel={t(
            'enrollment.button.cancelNotificationSubscription'
          )}
          subscribeLabel={t('enrollment.button.subscribeToNotifications')}
        />
      )}
      {submitType === SubmitTypes.externalTicketSystem && (
        <LinkButton to={eventUrl}>
          {t('event.register.occurrenceTableHeader.buttonText')}
        </LinkButton>
      )}
    </>
  );

  return (
    <>
      <tr className={[styles.occurrence, styles.isMobile].join(' ')}>
        <td>
          {occurrence.venue.name}{' '}
          <time dateTime={machineReadableDateTime}>
            {date} {time}
          </time>
        </td>
        {showFreePlaces && (
          <td className={styles.remainingCapacity}>{remainingCapacity}</td>
        )}
        <td className={styles.occurrenceSubmit}>
          {canEnroll ? submitCell : null}
        </td>
      </tr>
      <tr className={styles.occurrence}>
        <td>{date}</td>
        <td>{time}</td>
        <td>{occurrence.venue.name}</td>
        {showFreePlaces && (
          <td className={styles.remainingCapacity}>{remainingCapacity}</td>
        )}
        <td className={styles.occurrenceSubmit}>
          {canEnroll ? submitCell : null}
        </td>
      </tr>
    </>
  );
};

export default EventOccurrence;
