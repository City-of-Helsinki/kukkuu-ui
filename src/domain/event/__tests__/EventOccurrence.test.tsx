import React from 'react';

import {
  render,
  fireEvent,
  waitFor,
  selectAllHdsButtonByText,
  screen,
} from '../../../common/test/testingLibraryUtils';
import initModal from '../../../common/test/initModal';
import { eventQuery_event_occurrences_edges_node as OccurrenceEdgeNode } from '../../api/generatedTypes/eventQuery';
import EventOccurrence from '../EventOccurrence';

const mockedNode: OccurrenceEdgeNode = {
  id: 'T2NjdXJyZW5jZU5vZGU6Mg==',
  time: '2020-03-08T04:00:00+00:00',
  remainingCapacity: 99,
  event: {
    id: 'zzaaz',
    name: 'event name',
  },
  venue: {
    id: 'auppss',
    name: 'Musiikkitalo',
    address: '',
  },
  childHasFreeSpotNotificationSubscription: false,
};
const defaultProps = {
  occurrence: mockedNode,
};

const getWrapper = (props: unknown = {}) =>
  render(
    <table>
      <tbody>
        <EventOccurrence {...defaultProps} {...props} />
      </tbody>
    </table>
  );

it('renders snapshot correctly', () => {
  const { container } = getWrapper({ key: mockedNode.id });

  expect(container).toMatchSnapshot();
});

it('renders button for signing up to an event', () => {
  getWrapper();

  expect(screen.queryAllByText('Valitse')[0]).not.toEqual(null);
});

describe('when event is full', () => {
  const getFullEventWrapper = (
    occurrence: Record<string, unknown> | null = {},
    props: Record<string, unknown> = {}
  ) =>
    getWrapper({
      occurrence: {
        ...mockedNode,
        remainingCapacity: 0,
        ...occurrence,
      },
      ...props,
    });

  it('should show full label instead of remaining capacity', () => {
    getFullEventWrapper();

    expect(screen.queryAllByText('TÄYNNÄ')[0]).not.toEqual(null);
  });

  describe('and child has not subscribed to notifications', () => {
    it('user should be able to subscribe', async () => {
      initModal();

      const render = getFullEventWrapper(null);

      await waitFor(() => {
        fireEvent.click(selectAllHdsButtonByText(render, 'Tilaa ilmoitus'), {});
      });

      await waitFor(() => {
        // This button is hooked up to Apollo and mocks are not
        // provided, but this test still works.
        expect(screen.getAllByText('Tilaa ilmoitus')[0]).toBeDefined();
      });
    });
  });

  describe('and child has subscribed to notifications', () => {
    it('user should be able to unsubscribe', () => {
      getFullEventWrapper({
        childHasFreeSpotNotificationSubscription: true,
      });

      expect(screen.queryAllByText('Peru ilmoituksen tilaus')[0]).not.toEqual(
        null
      );
    });
  });
});
