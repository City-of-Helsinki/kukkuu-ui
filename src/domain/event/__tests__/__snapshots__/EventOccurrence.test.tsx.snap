// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<tr
  className="occurrence"
>
  <td
    className="occurrenceDate"
  >
    Su 3/8/2020
  </td>
  <td
    className="occurrenceTime"
  >
    06:00
  </td>
  <td
    className="occurrenceVenue"
  >
    Musiikkitalo
  </td>
  <td
    className="remainingCapacity"
  >
    99
  </td>
  <td
    className="occurrenceSubmit"
  >
    <Link
      to="zzaaz/occurrence/T2NjdXJyZW5jZU5vZGU6Mg==/enrol"
    >
      <Button
        className="submitButton"
        type="submit"
      >
        Valitse
      </Button>
    </Link>
  </td>
</tr>
`;
